File: Examples/SampleLocalzAttendantApp-ObjC/SampleLocalzAttendantApp-ObjC/ViewController.m
//
//  ViewController.m
//  SampleLocalzAttendantApp-ObjC
//
//  Created by Daniel Sykes-Turner on 13/2/18.
//  Copyright © 2018 Localz Pty Ltd. All rights reserved.
//

#import "ViewController.h"
#import <LocalzAttendantSDK/LocalzAttendantSDK.h>

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, weak) IBOutlet UITextField *txtOrderNumber;
@property (nonatomic, strong) NSMutableArray *orders;
@property (nonatomic, strong) NSString *attendantUsername;
@property (nonatomic, strong) NSString *attendantPassword;
@property (nonatomic, strong) NSString *storeBranchId;
@property (nonatomic, strong) NSString *storePickupId;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.attendantUsername = @"enter_attendant_username_here";
    self.attendantPassword = @"enter_attendant_password_here";
    self.storeBranchId = @"enter_store_branch_id_here";
    self.storePickupId = @"enter_store_pickup_id_here";
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    [[LocalzAttendantSDK shared] loginWithUsername:self.attendantUsername password:self.attendantPassword branchId:self.storeBranchId force:true options:nil completion:^(NSError * _Nullable error, LocalzAttendant * _Nullable user) {
        NSLog(@"Attendant logged in");
        [self reloadOrders];
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)reloadOrders {
    
    [[LocalzAttendantSDK shared] retrieveOrdersWithCompletion:^(NSError * _Nullable error, NSArray * _Nullable orders) {
        self.orders = orders.mutableCopy;
        [self.tableView reloadData];
    }];
}

#pragma mark - Order Flow

- (IBAction)createOrder:(id)sender {
    
    NSString *orderNumber = self.txtOrderNumber.text;
    if (orderNumber.length == 0) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"No Order Number" message:@"Please enter a number to create an order" preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alert animated:true completion:nil];
    } else {
        LocalzAttendantOrder *order = [[LocalzAttendantOrder alloc] initWithData:@{
                                                                             @"orderNumber":orderNumber,
                                                                             @"shopperId":@"customer1234",
                                                                             @"specific":@{
                                                                                     @"address":@"412 Collins Street, Melbourne, Victoria"
                                                                                     }
                                                                             }];
        
        [[LocalzAttendantSDK shared] createOrUpdateOrder:order completion:^(NSError * _Nullable error) {
            [self reloadOrders];
        }];
    }
    
    [self.txtOrderNumber resignFirstResponder];
}

- (void)checkinOrder:(NSString *)orderNumber {
    
    [[LocalzAttendantSDK shared] checkinOrderNumber:orderNumber pickupId:self.storePickupId subProjectId:nil completion:^(NSError * _Nullable error) {
        [self reloadOrders];
    }];
}

- (void)acknowledgeOrder:(NSString *)orderNumber {
    
    [[LocalzAttendantSDK shared] acknowledgeOrderNumber:orderNumber subProjectId:nil completion:^(NSError * _Nullable error) {
        [self reloadOrders];
    }];
}

- (void)completeOrder:(NSString *)orderNumber {
    
    [[LocalzAttendantSDK shared] completeOrderNumber:orderNumber signature:nil notes:nil subProjectId:nil completion:^(NSError * _Nullable error) {
        [self reloadOrders];
    }];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return self.orders.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"cell"];
    }
    
    LocalzAttendantOrder *order = self.orders[indexPath.row];
    cell.textLabel.text = order.orderNumber;
    cell.detailTextLabel.text = order.status;
    
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:true];
    
    LocalzAttendantOrder *order = self.orders[indexPath.row];
    if ([order.status isEqualToString:ORDER_STATUS_PENDING] || [order.status isEqualToString:ORDER_STATUS_READY]) {
        [self checkinOrder:order.orderNumber];
    } else if ([order.status isEqualToString:ORDER_STATUS_UNASSIGNED]) {
        [self acknowledgeOrder:order.orderNumber];
    } else if ([order.status isEqualToString:ORDER_STATUS_ASSIGNED]) {
        [self completeOrder:order.orderNumber];
    } else {
        NSLog(@"Unhandled order status: %@", order.status);
    }
}

@end

File: Examples/SampleLocalzAttendantApp-Swift/SampleLocalzAttendantApp-Swift/ViewController.swift
//
//  ViewController.swift
//  SampleLocalzAttendantApp-Swift
//
//  Created by Daniel Sykes-Turner on 13/2/18.
//  Copyright © 2018 Localz Pty Ltd. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var txtOrderNumber: UITextField!
    var orders: [LocalzAttendantOrder]! = [LocalzAttendantOrder]()
    
    let attendantUsername = "enter_attendant_username_here"
    let attendantPassword = "enter_attendant_password_here"
    let attendantBranchId = "enter_store_branch_id_here"
    let storePickUpId = "enter_store_pickup_id"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
        
        LocalzAttendantSDK.shared().login(withUsername: attendantUsername, password: attendantPassword, branchId: attendantBranchId, force: true, options: nil) { (error:Error?, attendant:LocalzAttendant?) in
            print("Attendant logged in")
            self.reloadOrders()
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func reloadOrders() {
        
        LocalzAttendantSDK.shared().retrieveOrders { (error:Error?, orders:[Any]?) in
            if let orders = orders as? [LocalzAttendantOrder] {
                self.orders = orders
            }
            self.tableView.reloadData()
        }
    }
    
    // MARK: - Order Flow
    
    @IBAction func createOrder() {
        
        let orderNumber:String! = self.txtOrderNumber.text
        if orderNumber == "" {
            let alert = UIAlertController.init(title: "No Order Number", message: "Please enter a number to create an order", preferredStyle: UIAlertControllerStyle.alert)
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
            self.present(alert, animated: true, completion: nil)
        } else {
            let order:LocalzAttendantOrder! = LocalzAttendantOrder.init(data: [
                "orderNumber":orderNumber,
                "shopperId":"customer1234",
                "specific":[
                    "address":"412 Collins St, Melbourne, Victoria"
                ]
                ])
            
            LocalzAttendantSDK.shared().createOrUpdate(order) { (error:Error?) in
                self.reloadOrders()
            }
        }
        
        self.txtOrderNumber.resignFirstResponder()
    }
    
    func checkInOrder(orderNumber: String!) {
        
        LocalzAttendantSDK.shared().checkinOrderNumber(orderNumber, pickupId: storePickUpId, subProjectId: nil) { (error:Error?) in
            self.reloadOrders()
        }
    }
    
    func acknowledgeOrder(orderNumber: String!) {
        
        LocalzAttendantSDK.shared().acknowledgeOrderNumber(orderNumber, subProjectId: nil) { (error:Error?) in
            self.reloadOrders()
        }
    }
    
    func completeOrder(orderNumber: String!) {
        
        LocalzAttendantSDK.shared().completeOrderNumber(orderNumber, signature: nil, notes: nil, subProjectId: nil) { (error:Error?) in
            self.reloadOrders()
        }
    }
    
    // MARK: - UITableViewDataSource
    
    func numberOfSections(in tableView: UITableView) -> Int {
        
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return self.orders.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCell(withIdentifier: "cell")
        if cell == nil {
            cell = UITableViewCell.init(style: UITableViewCellStyle.subtitle, reuseIdentifier: "cell")
        }
        
        let order = self.orders[indexPath.row]
        cell?.textLabel?.text = order.orderNumber
        cell?.detailTextLabel?.text = order.status
        
        return cell!
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        let order = self.orders[indexPath.row]
        if order.status == ORDER_STATUS_PENDING || order.status == ORDER_STATUS_READY {
            self.checkInOrder(orderNumber: order.orderNumber)
        } else if order.status == ORDER_STATUS_UNASSIGNED {
            self.acknowledgeOrder(orderNumber: order.orderNumber)
        } else if order.status == ORDER_STATUS_ASSIGNED {
            self.completeOrder(orderNumber: order.orderNumber)
        } else {
            print("Unhandled order status: \(order.status)")
        }
    }
}

